### ISSUE TITLE
compilation takes a long time in android studio

### ISSUE THREAD

**Issue by OlegAntipov** [Kode/KodeStudio#51]
Perhaps it is worth not to overwrite the project files for android native target, if they were not affected by the changes? I believe this would significantly increase the compilation speed.

Because even with the slightest change it is necessary to wait about 3-4 minutes in the android studio while the gradle syncs the project and compiles 500+ files.

Also, btw, somewhere can I specify which version of the gradle and build tools will be used in the project?

**Comment by RobDangerous** [comment]
Oh, did you find a workaround? Kode Studio always uses default parameters but when you compile from the command line you can use the --noproject parameter.

**Comment by OlegAntipov** [comment]
yes, I figured out what to do with this, but it's hard to call this method good) 

Below I'll write what I did, maybe someone will find it useful, because my compile time was reduced from 5-6 minutes to 30 seconds, and compiler rebuild ~14 files, not 500+. I use Android native build.

Here's what I did:
1) Update gradle and buildTools versions here KodeStudio\resources\app\extensions\kha\Kha\Kore\Tools\koremake\Data\android\build.gradle
KodeStudio\resources\app\extensions\kha\Kha\Kore\Tools\koremake\Data\android\app\build.gradle

_(I think it would be very cool to update this versions in khafile.js parameters)_

2) Most important thing: prevent overwriting of CMakeLists.txt if this file has not changed (as I understood, gradle starts re-indexing all the project completely if this file has different timestamp):

in KodeStudio\resources\app\extensions\kha\Kha\Kore\Tools\koremake\out\Exporters\AndroidExporter.js

replace
`fs.writeFileSync(path.join(outdir, 'app', 'CMakeLists.txt'), cmake, { encoding: 'utf8' });`

by that
```
		//compare
                var oldPath=path.join(outdir, 'app', 'CMakeLists.txt');
                var newPath=path.join(outdir, 'app', 'CMakeLists_new.txt');
		if(!fs.existsSync(oldPath))
			fs.writeFileSync(oldPath, cmake, { encoding: 'utf8' });
		else
		{
			fs.writeFileSync(newPath, cmake, { encoding: 'utf8' });
			
			var file1=fs.readFileSync(oldPath, 'utf-8');
			var file2=fs.readFileSync(newPath, 'utf-8');
			
			if(file1===file2)
				fs.unlinkSync(newPath);
			else
			{
				fs.unlinkSync(oldPath);
				fs.renameSync(newPath,oldPath);
			}
		}
		//end compare
```

3) A few tuning parameters that I believe speed up the build time:

in Android Studio:
File->Build,Execution,Deployment->Compiler
add in command line "--offline"

File->Build,Execution,Deployment->Gradle
check "Offline Work"

in KodeStudio\resources\app\extensions\kha\Kha\Kore\Tools\koremake\Data\android\app\build.gradle
```

android {

    lintOptions {
       checkAllWarnings false
       checkReleaseBuilds false
       ignoreWarnings true       // false by default
       quiet true                // false by default
    }
	
   //because I don't want sign each time manually
    signingConfigs {
        config {
            keyAlias '<your alias>'
            keyPassword '<your pass>'
            storeFile file('<path to *.jks>')
            storePassword '<your pass>'
        }
    }

....

    buildTypes {
        release {
            ......
            signingConfig signingConfigs.config
        }

        releaseArm7 {
            ....
	    signingConfig signingConfigs.config
        }

        debugArm7 {
            ......
	    signingConfig signingConfigs.config
        }
    }

```

in KodeStudio\resources\app\extensions\kha\Kha\Kore\Tools\koremake\Data\android\gradle.properties

```
org.gradle.jvmargs=-Xmx4096m -XX:MaxPermSize=1024m -XX:+HeapDumpOnOutOfMemoryError -Dfile.encoding=UTF-8
org.gradle.daemon=true
org.gradle.parallel=true
org.gradle.configureondemand=true
```

**Comment by RobDangerous** [comment]
Thanks, nice hints, I'll incorporate these.

**Comment by RblSb** [comment]
@OlegAntipov can be closed now? (Signing stuff here Kode/koremake/pull/51)

**Comment by RobDangerous** [comment]
I guess so.
